containers/ftp/ftp_files.rb

```diff
@@ -0,0 +1,208 @@
+          Timeout.timeout(60) do
+            ftp.connect(remote_host, remote_port)
+            ftp.login(remote_user, remote_password) if uri.userinfo
+          end
+        rescue StandardError => err
+          logger.error "FTP connect failed: #{cstr}: #{err.message}"
+          raise err
+        end
+
+        # if there is a remote dir that is not "."
+        if remote_directory && remote_directory != '.'
+          begin
+            Timeout.timeout(60) do
+              begin
+                ftp.mkdir(remote_directory)
+              rescue StandardError => err
+                logger.warn("FTP mkdir failed: #{cstr}: #{err.message}")
+              end
+              ftp.chdir(remote_directory)
+            end
+          rescue StandardError => err
+            logger.error "FTP chdir failed: #{cstr}: #{err.message}"
+            raise err
+          end
+        end
+
+        # deliver the file, catch errors and log when they occur
+        # give each file 1/2 hour to get ftp'd
+        begin
+          Timeout.timeout(1800) do
+            logger.debug("FTP start #{local_file.path} -> #{remote_file_name}")
+
+            last_noop = Time.now.to_i
+
+            # local_file_size = File.size(local_file)
+            bytes_uploaded = 0
+
+            ftp.put(local_file.path, remote_file_name) do |chunk|
+              bytes_uploaded += chunk.size
+
+              if (keep_alive > 0) && ((last_noop + keep_alive) < Time.now.to_i)
+                last_noop = Time.now.to_i
+
+                # this is to act as a keep alive - wbur needed it for remix delivery
+                begin
+                  ftp.noop
+                rescue StandardError => err
+                  # if they don't support this, and throw an error just keep going.
+                  logger.warn("FTP noop error, off and retry: #{err.message}")
+                  retry_count = [(retry_count - 1), 0].max
+                  keep_alive = 0
+                  raise err
+                end
+
+                # logger.debug "ftp put to #{remote_host}, #{bytes_uploaded} of #{local_file_size} bytes, (#{bytes_uploaded * 100 / local_file_size}%)."
+              end
+            end
+
+            logger.debug "FTP put #{local_file.path} as #{remote_file_name}"
+
+            if md5
+              ftp.puttextfile(md5_file.path, remote_file_name + '.md5')
+              logger.debug "FTP put #{md5_file.path} as #{remote_file_name}.md5"
+            end
+          end
+        rescue StandardError => ex
+          logger.error "FTP failed from '#{local_file.path}' to '#{cstr}'\n#{ex.message}\n\t" + ex.backtrace[0, 3].join("\n\t")
+          raise ex
+        end
+
+        # if we get here, then it all worked!
+        result = true
+
+        # this records success!
+        recorder.record('FtpSuccess', 'Count', 1.0)
+
+      rescue StandardError => err
+
+        # this records retried fails - open, login, mkdir, or put
+        recorder.record('FtpError', 'Count', 1.0)
```

Comment:
How many different FTP servers are we generally sending things to? Or I guess more importantly, how many are problematic and require retries? If it's not hundreds or thousands, would it be handy to capture the hostname as an additional dimension on the metric without getting too expensive?

Reply:
I'm not sure how many - I'd have to go look in exchange, between 10-100 or so I would estimate